src/main/java/com/yelp/nrtsearch/server/handler/SearchHandler.java

```diff
@@ -810,60 +809,61 @@ private CompositeFieldValue getFieldForHit(
-            };
-        DoubleValues doubleValues = virtualFieldDef.getValuesSource().getValues(leaf, scoreValue);
-        doubleValues.advanceExact(docID);
-        compositeFieldValue.addFieldValue(
-            FieldValue.newBuilder().setDoubleValue(doubleValues.doubleValue()));
-      } else if (fd instanceof RuntimeFieldDef) {
-        RuntimeFieldDef runtimeFieldDef = (RuntimeFieldDef) fd;
-        RuntimeScript.SegmentFactory segmentFactory = runtimeFieldDef.getSegmentFactory();
-        RuntimeScript values = segmentFactory.newInstance(leaf);
-        int docID = hit.getLuceneDocId() - leaf.docBase;
-        // Check if the value is available for the current document
-        if (values != null) {
-          values.setDocId(docID);
-          Object obj = values.execute();
-          ObjectToCompositeFieldTransformer.enrichCompositeField(obj, compositeFieldValue);
+      switch (fd) {
+        case VirtualFieldDef virtualFieldDef -> {
+          int docID = hit.getLuceneDocId() - leaf.docBase;
+
+          assert !Double.isNaN(hit.getScore()) || !virtualFieldDef.getValuesSource().needsScores();
+          DoubleValues scoreValue =
+              new DoubleValues() {
+                @Override
+                public double doubleValue() throws IOException {
+                  return hit.getScore();
+                }
+
+                @Override
+                public boolean advanceExact(int doc) throws IOException {
+                  return !Double.isNaN(hit.getScore());
+                }
+              };
+          DoubleValues doubleValues = virtualFieldDef.getValuesSource().getValues(leaf, scoreValue);
+          doubleValues.advanceExact(docID);
+          compositeFieldValue.addFieldValue(
+              FieldValue.newBuilder().setDoubleValue(doubleValues.doubleValue()));
         }
-
-      } else if (fd instanceof IndexableFieldDef && ((IndexableFieldDef) fd).hasDocValues()) {
-        int docID = hit.getLuceneDocId() - leaf.docBase;
-        // it may be possible to cache this if there are multiple hits in the same segment
-        LoadedDocValues<?> docValues = ((IndexableFieldDef) fd).getDocValues(leaf);
-        docValues.setDocId(docID);
-        for (int i = 0; i < docValues.size(); ++i) {
-          compositeFieldValue.addFieldValue(docValues.toFieldValue(i));
+        case RuntimeFieldDef runtimeFieldDef -> {
+          RuntimeScript.SegmentFactory segmentFactory = runtimeFieldDef.getSegmentFactory();
+          RuntimeScript values = segmentFactory.newInstance(leaf);
+          int docID = hit.getLuceneDocId() - leaf.docBase;
+          // Check if the value is available for the current document
+          if (values != null) {
+            values.setDocId(docID);
+            Object obj = values.execute();
+            ObjectToCompositeFieldTransformer.enrichCompositeField(obj, compositeFieldValue);
+          }
+        }
+        case IndexableFieldDef<?> fieldDef when fieldDef.hasDocValues() -> {
```

Comment:
TIL switch expression doesn't need explicit instanceof, and also the matching with `when`. I'm finding these large case blocks a bit odd, but I guess I can get used to these.